### PR TITLE
Maintenance

### PR DESCRIPTION
**Pull request by pieterbreed**

### CODE REVIEW

File: src/afrolabs/components/kafka.clj [hunk 1]
```diff
@@ -145,17 +145,29 @@
...
+                                             :partition (.partition ^RecordMetadata meta-data)}
+
+                                      (.hasOffset ^RecordMetadata meta-data)
+                                      (assoc :offset (.offset ^RecordMetadata meta-data))
+
+                                      (.hasTimestamp ^RecordMetadata meta-data)
+                                      (assoc :timestamp (t/instant (.timestamp ^RecordMetadata meta-data))))))
         (csp/close! delivered-ch)
-        (trace "onCompletion done.")))))
+        (trace "onCompletion delivered result.")))))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:170]: We enhance the capabilities of the _produce_ record's `:delivered-ch` channel. This channel used to received failures or exceptions (a type of failure) or in the case of success, the original record with `{:topic ... :value ...}` data. 

After this change
- more exceptions (failures) will be sent into the channel, 
- and the channel will be closed. 
- There was a way for the channel to remain open before.
- New fields are now available `:partition` & `:offset`. 
- This allows for creation of _Consistency Token_ values.

File: src/afrolabs/components/kafka.clj [hunk 2]
```diff
@@ -1547,6 +1559,74 @@
                            (s/keys :req-un [::ktable-id]
                                    :opt-un [::caught-up-once?])))
 
+(defprotocol IKTable
+  (ktable-wait-for-catchup
+    [_this topic-partition-offset timeout-period timeout-value]
+    "Waits until a ktable has consumed messages from its source topics, at least up to the topic-partition-offset data parameter.
+Supports a timeout operation."))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1566]: New protocol for an old concept.

File: src/afrolabs/components/kafka.clj [hunk 3]
```diff
@@ -1547,6 +1559,74 @@
...
+
+(defn ktable-atom-wait-for-catchup
+  "Will use csp to actually wait up to timeout-duration for the ktable value to reflect changes up to this level."
+  [ktable-atom topic-partition-offset timeout-duration timeout-value]
+  (let [timeout-ms (.toMillis ^java.time.Duration timeout-duration)
+        timeout-chan (csp/timeout timeout-ms)
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1572]: This provides the upper bound for how long we will wait for a result. This channel will close after the `timeout-ms` ms ellapsed.

File: src/afrolabs/components/kafka.clj [hunk 4]
```diff
@@ -1547,6 +1559,74 @@
...
+                                    (keep (fn [[topic partition offset]]
+                                            (let [ktable-progress-offset (get-in ktable-topic-partition-offsets
+                                                                                 [topic partition])]
+                                              (when (< ktable-progress-offset offset)
+                                                :not-caught-up-yet))))
+                                    (count)
+                                    (zero?))))]
+    (cond
+      (caught-up-already? initial-ktable-value)
+      (ktable->topic-partition-offsets initial-ktable-value)
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1596]: Early return if we start out knowing about those topics-partition-offsets already.

File: src/afrolabs/components/kafka.clj [hunk 5]
```diff
@@ -1547,6 +1559,74 @@
...
+
+      :else
+      (let [watch-id (keyword "ktable-atom-wait-for-catchup" (str (random-uuid)))
+            caught-up?-chan (csp/chan)
+            new-ktable-ch (csp/chan (csp/sliding-buffer 1))]
+
+        ;; notice changes when they happen
+        (add-watch ktable-atom watch-id (fn [_ _ _ _] (csp/>!! new-ktable-ch true)))
+
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1605]: Use a watch mechanism to _know about_ updates, which triggers a check.

File: src/afrolabs/components/kafka.clj [hunk 6]
```diff
@@ -1547,6 +1559,74 @@
...
+              (do (csp/>! caught-up?-chan timeout-value)
+                  (csp/close! caught-up?-chan)
+                  timeout-value)
+              (let [ktable-value @ktable-atom]
+                (if (caught-up-already? ktable-value)
+                  (do (csp/>! caught-up?-chan (ktable->topic-partition-offsets ktable-value))
+                      (csp/close! caught-up?-chan))
+                  (recur (csp/alts! [new-ktable-ch
+                                     timeout-chan
+                                     (csp/timeout (long (/ timeout-ms 10)))]))))))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1620]: on a go-thread:
- wait only as long as the outside `timeout-chan` channel gives us
- if a trigger arrives to check, we take a value of the atom, and checks if it has caught up yet or not.
- also check regularly

File: src/afrolabs/config.clj [hunk 7]
```diff
@@ -104,6 +104,22 @@
...
+  [value]
+  (boolean (cond
+             (nil? value)     nil ;; (boolean nil) -> false
+             (boolean? value) value
+             ;; environment variables come as strings.
+             (string? value)  (#{"ON" "1" "TRUE"}     ;; any envvar value not exactly these ones will be false
+                               (str/upper-case value))
+             (pos-int? value) true
+             (zero?    value) false
+             :else            value)))
```
Comment (pieterbreed) [inline src/afrolabs/config.clj:117]: Turns strings such as `"1" "ON" "True" "TRUE"` into `true` and strings that are something else into `false`.

**Review by pieterbreed — COMMENTED** [review]
Author's notes

**Review by BenjaminAfrolabs — APPROVED** [review]
Looks good from what I can read :D

**Review by faf-gm — APPROVED** [review]

**Review by faf-gm — APPROVED** [review]

**Pull request merged**